File: u1qns/Week_21/PGS_산 모양 타일링.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(int n, vector<int> tops) {
    int answer = 0;
    vector< vector<int> > dp(2, vector<int> (n, 0));


    // 탑의 유무로 만들지 못하는 경우가 생김
    dp[0][1] = 2;
    dp[1][1] = 3;
    for(int i=2; i<n; ++i)
    {
        if(tops[i-1] == 1)
        {
            // 직전에 탑이 있으면 모양을 더 만들 수 있지만
            // 타일이 서로 겹치는 부분이 생기며 이 부분을 빼야함 
        }
        else
        {
          // 직전 탑이 없으면..
        }
            
        
        //cout << dp[0][i] << "\t" << dp[1][i] << "\n";
    }
    
}

File: seoyoung059/Week_21/PGS_산_모양_타일링/PGS_산_모양_타일링.java
package Week_21.PGS_산_모양_타일링;

class PGS_산_모양_타일링 {
    public int solution(int n, int[] tops) {
        int answer = 0;
        int[][] dp = new int[n + 1][2];
        dp[0][0] = 1;

        for (int i = 1; i <= n; i++) {
            if (tops[i - 1] == 1) {
                dp[i][0] = (dp[i - 1][0] * 3 + dp[i - 1][1] * 2) % 10007;
                dp[i][1] = (dp[i - 1][0] + dp[i - 1][1]) % 10007;
            } else {
                dp[i][0] = (dp[i - 1][0] * 2 + dp[i - 1][1]) % 10007;
                dp[i][1] = (dp[i - 1][0] + dp[i - 1][1]) % 10007;
            }
        }

        answer = (dp[n][0] + dp[n][1]) % 10007;
        return answer;
    }
}
